src/rnn.cpp

```diff
@@ -897,80 +985,595 @@ void RNNDescriptor::GetLayerBiasOffset(const int layer,
 
-            const std::vector<size_t> copy_size{static_cast<size_t>(copy_seq_cnt),
-                                                static_cast<size_t>(copy_bsize),
-                                                static_cast<size_t>(vector_size)};
+            bool layout_equal =
+                (desc_seq_is_padded == base_seq_is_padded) &&
+                std::equal(desc_dim_order.begin(), desc_dim_order.end(), base_dim_order.begin());
 
-            auto packed_desc = miopen::TensorDescriptor(padded_tensor_desc.GetType(), copy_size);
-            auto padded_desc =
-                miopen::TensorDescriptor(padded_tensor_desc.GetType(), copy_size, padded_stride);
+            if(layout_equal)
+                return base_layout;
+        }
+    }
+    return miopenRNNDataUnknownLayout;
+}
 
-            // Result from GetElementSpace does not include the padding from the last sequence
-            // WA: So calculated manually.
-            unsigned int WA_padded_ElementSpace = padded_stride[0] * copy_size[0];
+SeqTensorDescriptor RNNDescriptor::makeSeqTensorDescriptor(miopenDataType_t t,
+                                                           miopenRNNBaseLayout_t layout,
+                                                           int maxSeqLength,
+                                                           int batchSize,
+                                                           int vectorSize,
+                                                           const int* lensPerSeq)
+{
+    const std::vector<int> lens = {batchSize, maxSeqLength, vectorSize};
 
-            if(is_src_padded)
-            {
-                CopyTensor(
-                    handle, padded_desc, src, packed_desc, dst, src_offset, dst_offset, true);
+    const auto [dim_order, padded_sequences] = convertRNNBaseLayout(layout);
 
-                src_offset += WA_padded_ElementSpace;
-                dst_offset += packed_desc.GetElementSpace();
-            }
-            else
-            {
-                CopyTensor(
-                    handle, packed_desc, src, padded_desc, dst, src_offset, dst_offset, true);
+    return {t,
+            dim_order,
+            lens,
+            std::vector<int>(lensPerSeq, lensPerSeq + batchSize),
+            true,
+            padded_sequences};
+}
 
-                src_offset += packed_desc.GetElementSpace();
-                dst_offset += WA_padded_ElementSpace;
-            }
-            left_id = i;
+void RNNDescriptor::SeqTensorToTensorDescArray(const SeqTensorDescriptor& desc,
+                                               std::vector<miopen::TensorDescriptor>& td,
+                                               std::vector<miopenTensorDescriptor_t>& ptd)
+{
+    if(!desc.IsPacked())
+        MIOPEN_THROW(miopenStatusInternalError, "Only packed SeqTensorDescriptor supported.");
+
+    const std::vector<size_t> bs = desc.GetBatchesPerSequence();
```

Comment:
Should it be copied?
```suggestion
    const std::vector<size_t>& bs = desc.GetBatchesPerSequence();
```

Reply:
this is not a direct GeterFunction. This should take calss field Seq(perBatch) and convert it to BatchesPerSequence